tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/RecordingZ3SolverContext.scala

```diff
@@ -0,0 +1,207 @@
+    val context = new RecordingZ3SolverContext(parentLog, config)
+    parentLog match {
+      case Some(log) =>
+        log.replay(context.solver)
+
+      case None => ()
+    }
+    context
+  }
+}
+
+/**
+  * A wrapper around SolverContext that records SMT operations. The log is organized hierarchically,
+  * so a differential diff may be taken.
+  *
+  * @param parentLog the log of the parent context
+  * @param config solver configuration
+  *
+  * @author Igor Konnov
+  */
+@SerialVersionUID(700L)
+class RecordingZ3SolverContext private (parentLog: Option[SmtLog], val config: SolverConfig)
+    extends SolverContext with Serializable with LazyLogging {
+
+  import SmtLog._
+
+  private val solver = new Z3SolverContext(config)
+
+  /**
+    * The sequence of logs (the last added element is in the head),
+    * one per context, except the last log, which is maintained in lastLog.
+    * Every list in logStackRev is stored in the reverse order.
+    */
+  private var logStackRev: List[List[Record]] = List()
+
+  /**
+    * The current log, the last added element is in the head.
+    */
+  private var lastLogRev: List[Record] = List()
+
+  /**
+    * Extract a differential SMT log from the current context.
+    * @return SMT log
+    */
+  def extractLog(): SmtLog = {
+    val newRecords = (lastLogRev ++ logStackRev.flatten).reverse
+    new SmtLog(parentLog, newRecords)
+  }
+
+  /**
+    * Get the current context level, that is the difference between the number of pushes and pops made so far.
+    *
+    * @return the current level, always non-negative.
+    */
+  override def contextLevel: Int = solver.contextLevel
+
+  /**
+    * Save the current context and push it on the stack for a later recovery with pop.
+    */
+  override def push(): Unit = {
+    solver.push()
+    logStackRev = lastLogRev :: logStackRev
+    lastLogRev = List()
+  }
+
+  /**
+    * Pop the previously saved context. Importantly, pop may be called multiple times and thus it is not sufficient
+    * to save only the latest context.
+    */
+  override def pop(): Unit = {
+    solver.pop()
+    lastLogRev = logStackRev.head
+    logStackRev = logStackRev.tail
+  }
+
+  /**
+    * Pop the context as many times as needed to reach a given level.
+    *
+    * @param n pop n times, if n > 0, otherwise, do nothing
+    */
+  override def pop(n: Int): Unit = {
+    solver.pop(n)
+    logStackRev = logStackRev.drop(n - 1) // n - 1 in logStack + 1 in lastLog
+    lastLogRev = logStackRev.head
+    logStackRev = logStackRev.tail
+  }
+
+  /**
+    * Clean the context
+    */
+  override def dispose(): Unit = {
+    solver.dispose()
+  }
+
+  /**
+    * Declare a constant for an arena cell.
+    * This method is called automatically by the arena.
+    *
+    * @param cell a (previously undeclared) cell
+    */
+  override def declareCell(cell: ArenaCell): Unit = {
+    lastLogRev = DeclareCellRecord(cell) :: lastLogRev
```

Comment:
Seems like it might be cleaner to abstract this mutable stack pushing by adding a method like `push(r: Record): Unit` that took care of the consing and mutation of `lastLogRev`.

Reply:
Yep. Added `appendToTopLog`.